i888 Use correct java compiler for JUnit tests

johnbrvc opened the pull request:
### Description of what the PR does
Since _Eclipse 4.17 (2020-09)_, **Java 11** (or newer) is the default JDK that ships with Eclipse.  Previously, it was **Java 8** (1.8). This PR attempts to track down the correct JDK to use for compiling the _JUnit_ test programs.  The JDK must match the currently running JVM in order to qualify.  A search of  the `JAVA_HOME/bin` and folders on the `PATH `environment variable are checked for the presence of the corresponding JDK (`javac[.exe]`). The code only checks for the presence of the Java version string in the folder name.  It could be expanded to actually attempt to run "`javac -version`" in each folder, but that shouldn't be necessary, and is highly inefficient.

The reason for this PR is that the JUnits requiring compiles were using Eclipse's JDK (Version >11), but attempting to execute the produced class files using JRE 1.8.  This lead to the errors described in #888.

### Issue which the PR addresses
Fixed #888 

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 11
Eclipse 2021-12, with JVM org.eclipse.justj.openjdk.hotspot.jre.full.win32.x86_64_17.0.1.v20211116-1657
JDK 1.8.0_351

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)
Run all JUnits.  They should pass now.   Previously, the ones in #888 (ExecuteTest, among others) did not pass.

johnbrvc committed:
i888 Use correct java compiler for JUnit tests

Since Eclipse 4.17 (2020-09), Java 11 is the default JDK that ships with Eclipse.  Previously, it was Java 8 (1.8).
This PR attempts to track down the correct JDK to use for compiling the JUnit test programs.  The JDK must match the currently running JVM in order to qualify.  A search of  the JAVA_HOME/bin and folders on the PATH environment variable are checked for the presence of the corresponding JDK (javac[.exe]).
The code only checks for the presence of the java version string in the folder name.  It could be expanded to actually attempt to run "javac -version" in each folder, but that shouldn't be necessary.

troy2914 commented:
Tested with jdk17 as first in my PATH (and only java in my PATH),  generated a `version 61.0` PracticeA.class
added my Amazon Correta 8 to the end my PATH /Library/Java/JavaVirtualMachines/amazon-corretto-8.jdk/Contents/Home/bin,
still generated a version 61.0 class.

troy2914 commented:
btw, standard location on MacOS is for the jdk to be under 
/Library/Java/JavaVirtualMachines
where I have
```
/Library/Java/JavaVirtualMachines/jdk1.8.0_144.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_151.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_171.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_181.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_201.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_45.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_51.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_60.jdk
/Library/Java/JavaVirtualMachines/jdk1.8.0_65.jdk
```  
along with the amazon-corretto one

johnbrvc commented:
Based on @troy2914 's comments, and what I do see on Linux (JDK not located in a folder that contains the version number), we may have to what I suggested in the PR comment, that is, execute "`javac -version`" to see what version each **javac** is.  I'm thinking perhaps if we make the lookup of **javac** _static_ and save it to a _static_ member variable, then we only have to find the right **javac** once per **JUnit** test.

johnbrvc committed:
i888 test result of javac -version to determine jdk version

In order to track down the correct JDK to use for building java applications as part of a JUnit, we try each folder of: JAVA_HOME/bin, and PATH components until we find one that matches the JRE.

johnbrvc commented:
> Tested with jdk17 as first in my PATH (and only java in my PATH), generated a `version 61.0` PracticeA.class added my Amazon Corretto 8 to the end my PATH /Library/Java/JavaVirtualMachines/amazon-corretto-8.jdk/Contents/Home/bin, still generated a version 61.0 class.

Changed so it actually runs "javac -version" on every javac or javac.exe on the JAVA_HOME/bin and PATH until it finds one that matches instead of just looking for the version in the folder string.

johnbrvc commented:
> Based on @troy2914 's comments, and what I do see on Linux (JDK not located in a folder that contains the version number), we may have to what I suggested in the PR comment, that is, execute "`javac -version`" to see what version each **javac** is. I'm thinking perhaps if we make the lookup of **javac** _static_ and save it to a _static_ member variable, then we only have to find the right **javac** once per **JUnit** test.

Implemented this.
